### DISCUSSION TITLE
Simplifiying code by adding functions

### DISCUSSION THREAD

**Discussion by pakricard** [timschofield/webERP#239] [Development Discussion & Specification]
Hi all:
As webERP project grew larger over the years, we ended up with many lines of repeated code here and there, some of them with inconsistencies.

One of the easy ways to keep the spirit of webERP coding standards and simplify the code is the use of functions to encapsulate repeated code. It helps maintain consistency and diminishes the number of bugs, as code maintenance is simpler.

As an example:
Select Product line 308 to 317
```
	$SQL = "SELECT SUM(qtypu*(woitems.qtyreqd - woitems.qtyrecd)) AS woqtydemo
			FROM woitems INNER JOIN worequirements
			ON woitems.stockid=worequirements.parentstockid
			INNER JOIN workorders
			ON woitems.wo=workorders.wo
			AND woitems.wo=worequirements.wo
			INNER JOIN locationusers ON locationusers.loccode=workorders.loccode AND locationusers.userid='" .  $_SESSION['UserID'] . "' AND locationusers.canview=1
			WHERE  worequirements.stockid='" . $StockID . "'
			AND workorders.closed=0";
	$ErrMsg = _('The workorder component demand for this product cannot be retrieved because');
	$DemandResult = DB_query($SQL, $ErrMsg);
	if (DB_num_rows($DemandResult) == 1) {
		$DemandRow = DB_fetch_row($DemandResult);
		$Demand+= $DemandRow[0];
	}
```

StockStatus.php lines 173 to 189

```
	$SQL = "SELECT SUM(qtypu*(woitems.qtyreqd - woitems.qtyrecd)) AS woqtydemo
			FROM woitems INNER JOIN worequirements
			ON woitems.stockid=worequirements.parentstockid
			INNER JOIN workorders
			ON woitems.wo=workorders.wo
			AND woitems.wo=worequirements.wo
			WHERE workorders.loccode='" . $MyRow['loccode'] . "'
			AND worequirements.stockid='" . $StockID . "'
			AND workorders.closed=0";

	$ErrMsg = _('The workorder component demand for this product from') . ' ' . $MyRow['loccode'] . ' ' . _('cannot be retrieved because');
	$DemandResult = DB_query($SQL,$ErrMsg,$DbgMsg);

	if (DB_num_rows($DemandResult)==1){
		$DemandRow = DB_fetch_row($DemandResult);
		$DemandQty += $DemandRow[0];
	}

```

These two bits of code should do the same, but are a bit inconsistent as one takes care of the locations available to the user, but the other doesn't.

We could create a 

```
function GetItemDemandAsAComponentOfWorkOrders($Stockid){
}
```

that encapsulates the code, reduces lines on the scripts and we still have the readability we all love. Whenever the Demand of an item as component needs some change, we only need to code and debug it once. 

Are we OK to simplify / reduce code this way?

P.S.: I applied this technique to my custom scripts and it really simplifies maintenance, and I'm happy with the result.
P.S.2: This is just a random exmaple of code. there are plenty more repeated bits of code here and there.

**Comment by timschofield** [comment]
I am fine with these changes if you want to send them as a pull request
Ricard

Thanks
tim

On Fri, 15 Nov 2024 at 02:25, pakricard ***@***.***> wrote:

> Hi all:
> As webERP project grew larger over the years, we ended up with many lines
> of repeated code here and there, some of them with inconsistencies.
>
> One of the easy ways to keep the spirit of webERP coding standards and
> simplify the code is the use of functions to encapsulate repeated code. It
> helps maintain consistency and diminishes the number of bugs, as code
> maintenance is simpler.
>
> As an example:
> Select Product line 308 to 317
>
> 	$SQL = "SELECT SUM(qtypu*(woitems.qtyreqd - woitems.qtyrecd)) AS woqtydemo
> 			FROM woitems INNER JOIN worequirements
> 			ON woitems.stockid=worequirements.parentstockid
> 			INNER JOIN workorders
> 			ON woitems.wo=workorders.wo
> 			AND woitems.wo=worequirements.wo
> 			INNER JOIN locationusers ON locationusers.loccode=workorders.loccode AND locationusers.userid='" .  $_SESSION['UserID'] . "' AND locationusers.canview=1
> 			WHERE  worequirements.stockid='" . $StockID . "'
> 			AND workorders.closed=0";
> 	$ErrMsg = _('The workorder component demand for this product cannot be retrieved because');
> 	$DemandResult = DB_query($SQL, $ErrMsg);
> 	if (DB_num_rows($DemandResult) == 1) {
> 		$DemandRow = DB_fetch_row($DemandResult);
> 		$Demand+= $DemandRow[0];
> 	}
>
> StockStatus.php lines 173 to 189
>
> 	$SQL = "SELECT SUM(qtypu*(woitems.qtyreqd - woitems.qtyrecd)) AS woqtydemo
> 			FROM woitems INNER JOIN worequirements
> 			ON woitems.stockid=worequirements.parentstockid
> 			INNER JOIN workorders
> 			ON woitems.wo=workorders.wo
> 			AND woitems.wo=worequirements.wo
> 			WHERE workorders.loccode='" . $MyRow['loccode'] . "'
> 			AND worequirements.stockid='" . $StockID . "'
> 			AND workorders.closed=0";
>
> 	$ErrMsg = _('The workorder component demand for this product from') . ' ' . $MyRow['loccode'] . ' ' . _('cannot be retrieved because');
> 	$DemandResult = DB_query($SQL,$ErrMsg,$DbgMsg);
>
> 	if (DB_num_rows($DemandResult)==1){
> 		$DemandRow = DB_fetch_row($DemandResult);
> 		$DemandQty += $DemandRow[0];
> 	}
>
>
> These two bits of code should do the same, but are a bit inconsistent as
> one takes care of the locations available to the user, but the other
> doesn't.
>
> We could create a
>
> function GetItemDemandAsAComponentOfWorkOrders($Stockid){
> }
>
> that encapsulates the code, reduces lines on the scripts and we still have
> the readability we all love. Whenever the Demand of an item as component
> needs some change, we only need to code and debug it once.
>
> Are we OK to simplify / reduce code this way?
>
> P.S.: I applied this technique to my custom scripts and it really
> simplifies maintenance, and I'm happy with the result.
> P.S.2: This is just a random exmaple of code. there are plenty more
> repeated bits of code here and there.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/timschofield/webERP/discussions/239>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAL6LEUFKOENNTL7MEBOUD2AVLRBAVCNFSM6AAAAABR2G3ODCVHI2DSMVQWIX3LMV43ERDJONRXK43TNFXW4OZXGQ4DKNJUG4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
www.weberp.org <http://www.weberpafrica.com>
@TimSchofield2
Blog: https://kwamoja.home.blog/

**Comment by pakricard** [comment]
Committed ;-)
